chore(manifest): add spot subfield

SoManyHs opened the pull request:
This adds a new field, `spot` to the manifest as a subfield of `count`.  

First step of https://github.com/aws/copilot-cli/issues/2162

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

SoManyHs committed:
feat: add spot fields to manifest

cristim commented:
Awesome to see this being worked on, looking forward to see this released. Is there an ETA for the next release that includes this?

In the meantime I'll just build from @SoManyHs's fargate-spot branch to give it a try

Lou1415926 commented on internal/pkg/manifest/svc.go:
```diff
@@ -73,18 +73,68 @@ func (a *Count) UnmarshalYAML(unmarshal func(interface{}) error) error {
+		}
+	}
+
+	if *s.Base != 0 {
```
I'm wondering if `s.Base != nil` would do? I remember `s.Base` would be `nil` if unmarshal fails, and we probably won't be able to dereference it as a result. Or is this needed because we don't want the users to put `0` for `spot`?

SoManyHs replied:
It does not -- I tried :( And yeah it's because the default value for a nil pointer is the zero value, which is 0 in this case.

iamhopaul123 replied:
Maybe this just to be safe (in case `yaml` changes it's behavior)
```suggestion
	if aws.StringValue(s.Base) != 0 {
```

Lou1415926 commented on internal/pkg/manifest/svc.go:
```diff
@@ -73,18 +73,68 @@ func (a *Count) UnmarshalYAML(unmarshal func(interface{}) error) error {
+		}
+	}
+
+	if s.Enabled != nil {
```
(If we need to use `if *s.Base != 0` as opposed to `if s.Base != nil` above, then please ignore this comment)

Would `s.Base`  have been `nil` anyway if we get to this point?

SoManyHs replied:
Without this block, I get this error:
```
$ go test ./internal/pkg/manifest/ -run TestCount_UnmarshalYAML
--- FAIL: TestCount_UnmarshalYAML (0.00s)
    --- FAIL: TestCount_UnmarshalYAML/With_spot_enabled_with_autoscaling_range (0.00s)
        svc_test.go:344: 
            	Error Trace:	svc_test.go:344
            	Error:      	Received unexpected error:
            	            	cannot unmarshal "spot" field to an integer or boolean
            	Test:       	TestCount_UnmarshalYAML/With_spot_enabled_with_autoscaling_range
FAIL
```

Lou1415926 commented on internal/pkg/manifest/svc.go:
```diff
@@ -73,18 +73,68 @@ func (a *Count) UnmarshalYAML(unmarshal func(interface{}) error) error {
+		return nil
+	}
+
+	if err := unmarshal(&s.Enabled); err != nil {
```
(If we need to use `if *s.Base != 0` as opposed to `if s.Base != nil` above, then please ignore this comment)

If we no longer need the block that follows (which sets `s.Base = nil`), then we could probably just `return unmarshal(&s.Enabled)`?

Lou1415926 commented on internal/pkg/manifest/svc.go:
```diff
@@ -73,18 +73,68 @@ func (a *Count) UnmarshalYAML(unmarshal func(interface{}) error) error {
+
+	if err := unmarshal(&s.Enabled); err != nil {
+		switch err.(type) {
+		case *yaml.TypeError:
```
Perhaps `err` doesn't need to be handled differently here. It seems like if we get to the point of trying to unmarshal `s.Enabled`, it would mean that we've failed the attempt for `s.Base`. Hence if `unmarshal(&s.Enabled)` still returns an `yaml.TypeError`, that'd imply something is wrong with the manifest input.

SoManyHs replied:
Yeah, that's what I had originally -- it was def tricky figuring out how to handle the yaml errors here since we're treating `spot` effectively as a union type -- I tried to follow the example for [BuildArgsOrString](https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/manifest/workload.go#L249) but for whatever reason the unit tests weren't behaving properly when I didn't have both error handling cases here. Part of it was also tied up in the fact that `spot` accepts an integer, but that was getting read as the value for `count` at one point, but I can try to dig further and see if I can refactor!

Lou1415926 reviewed (commented):
Thank you so much for your work ❤️  I have a few questions below, two of which can be ignored depending on the answer to the first question.

iamhopaul123 commented on internal/pkg/manifest/svc.go:
```diff
@@ -63,8 +63,8 @@ type Count struct {
 // This method implements the yaml.Unmarshaler (v2) interface.
-func (a *Count) UnmarshalYAML(unmarshal func(interface{}) error) error {
-	if err := unmarshal(&a.Autoscaling); err != nil {
+func (c *Count) UnmarshalYAML(unmarshal func(interface{}) error) error {
```
good catch!

iamhopaul123 reviewed (commented):
Thank you for taking the first initiative for this! +1 to wanxian's comment. And did you run `go mod clean` to see if the `go.sum` changes go away, since there seems to be no import change? Also we might not want to include changes for the `package-lock.json` file 🤔

SoManyHs committed:
refactor: custom unmarshaller on spot type

SoManyHs committed:
chore: change spot field to be just int

Previously, spot was a custom type that could be either an int or bool
ptr. This eliminates the need for using a type union and allows
customers to specify desiredCount for their services under the `spot`
subfield to place all services onto a Fargate Spot capacitity provider.

iamhopaul123 reviewed (approved):
LGTM :shipit:

efekarakus reviewed (approved):
:shipit:
